application/org.openjdk.jmc.flightrecorder.flameview/src/main/java/org/openjdk/jmc/flightrecorder/flameview/views/FlameGraphView.java

```diff
@@ -382,19 +387,21 @@ public void selectionChanged(IWorkbenchPart part, ISelection selection) {
 
 	private void triggerRebuildTask(IItemCollection items) {
 		// Release old model calculation before building a new
-		if (modelCalculationFuture != null) {
-			modelRebuildCallable.setInvalid();
-			modelCalculationFuture.cancel(true);
+		if (!modelRebuildRunnableList.isEmpty()) {
+			modelRebuildRunnableList.get(activeModelRebuildThread).setInvalid();
+			activeModelRebuildThread = (activeModelRebuildThread + 1) % MODEL_EXECUTOR_THREADS_NUMBER;
```

Comment:
You don't need to keep track of the threads, do you? You just need to keep track of the latest runnable. If you get another request, you simply invalidate the current one (and then you don't care what happens with it, it will worst case continue for a little bit, and then fizzle out), and update the latest to be the one from the newest request.

I.e. the only reason we're having multiple threads, it so that we allow the discovery of the cancellation to take a little bit of time, during which we can already get going with the next one. There is only ever one calculation that will have a result we care about, and that is the latest one submitted.

Reply:
correct, we don't need a references as the only valid one is the last one. We have stages and only the last active sample set is displayed even when other threads end properly. The thread number is equal stages as during a tests it got the best UI experience working with large samples set.